### ISSUE TITLE
Push notification are not sending out

### ISSUE THREAD

**Issue by khantthulinn** [jleclanche/django-push-notifications#332]
I have tested with 3 cases. 

1) From admin module, I use 'Send test message'
2) I write my own view and send message to all 
```

class PushNotification(APIView):
    permission_classes = (IsAuthenticated,)

    def post(self, request):
    
        data = request.data
        message = data['message']
        devices = APNSDevice.objects.filter()
        devices.send_message(message)
        
        return Response(status=status.HTTP_200_OK)

```

3) I write my own view and send message to particular token

```

class PushNotificationToSpecificToken(APIView):
    permission_classes = (IsAuthenticated,)

    def post(self, request):
    
        data = request.data
        message = data['message']
        token = data['token']
        device = APNSDevice.objects.get(registration_id=token)
        device.send_message(message)
        
        return Response(status=status.HTTP_200_OK)
```

1 and 3 are still okay. It means my push certificate is working well. Problem is that when I send message to all user, it doesn't send already. Last time, I only have like 10 devices and it is okay. Currently, I am testing with 100 devices and it is not okay already. How shall I do?

**Comment by jamaalscarlett** [comment]
Are you receiving any errors?  When you use the admin can you bulk send to all devices?

**Comment by khantthulinn** [comment]
How can I check error message? According to my code, I will always get HTTP_200_OK. I haven't tried for bulk yet. Currently, it is already in production server and I don't have dev server yet.

**Comment by jamaalscarlett** [comment]
@khantthulinn if it sends for 10 devices, but not 100 devices, that sounds like a problem with the APNS service and not this library.

**Comment by andhieka** [comment]
@khantthulinn I just confirmed that on my server, method 2 is still working. However, yesterday I experienced a similar problem: the bulk message is not sending sometimes. It turns out that when there is a mixture of devices with Sandbox token and Production token, the notification will not get pushed to devices by APNS. Maybe you can do that check and try "Test bulk notification" on a subset of your devices? Just my two cents.

**Comment by khantthulinn** [comment]
@andhieka
Yes. Exactly same as my issue. I am using both sandbox and production token. When I need to send to both lite and prod, I change my cert manually. 

I have tested for all user (around 100) with "Test bulk notification". It is okay. 

It is not okay only when I use my own view. How shall I do?

**Comment by andhieka** [comment]
My suggestion is to try isolating the problem first. Go to Django shell (python manage.py shell) and test these lines:

from push_notifications.models import *

message = 'test message'
devices = APNSDevice.objects.filter()
devices.send_message(message)

Does this run ok? If it's ok, then the problem is not in these lines. If this is not ok, then find out first what works. (Using objects.all() instead of objects.filter()? Or trying objects.all().order_by('id') maybe). My suggestions might seem absurd but those are the things I tried myself.

**Comment by khantthulinn** [comment]
Thanks. After I test, I will come back to you. 

@jamaalscarlett i will check first.

**Comment by pfertyk** [comment]
Hey, I think I have a similar problem. There are 3 APNS devices shown on django admin page for my user:
- iPhone1 (correct)
- incorrect device (created manually by me by copying the data from another entry and modifying the **Registration ID**)
- iPhone2 (correct)

Both iPhones are separate physical devices. When I select 'Send test message' from django admin page (all 3 devices are selected) then both iPhones receive the message. However, when I select 'Send test message in bulk' (again, all 3 devices are selected), only iPhone1 receives the message. If I select only the two correct devices or I set **Is active** to `false` for the incorrect device, then both iPhones receive the bulk message correctly.

It seem that `django-push-notifications` are trying to send bulk messages in order the devices appear on django admin page and if any error occurs the whole process stops. Please let me know if this issue is related the one reported by @khantthulinn and what can be done to fix it? Is there an option for `apns_send_bulk_message` to ignore the errors or do I have to use the `for` loop each time I send multiple notifications? Or maybe I can somehow detect the incorrect registration ids and remove them from the database before I send the notifications?

**Comment by shigmas** [comment]
Last I read (which was a while ago), that is the expected behavior according to the documentation. The feedback service can be used to find invalid devices since the last notification. (Does that still work?)

**Comment by pfertyk** [comment]
So the solution is to send a message to each of the devices first and check if it was delivered, and if not, delete the device? Or do I get it totally wrong?

This is my first encounter with APNS, so I ask for patience :) Could you please provide a code sample that would illustrate the correct solution?

**Comment by r0y3** [comment]
Has anybody found a solution on this issue?

My experience is that when use the when select all devices on admin and select "Send test message in bulk" the iOS devices can receive a push notification message. But if it's triggered on command line or cron job the iOS devices can't receive any push notifications.

Thanks,

**Comment by gabreho** [comment]
@royge I'm having the same issue, how did you solve it?

**Comment by royge** [comment]
@gabreho I instead send the message individually rather than in bulk because if there is an inactive device  in the batch all other active are affected by the error. That is my cheap work-around :).

**Comment by jamaalscarlett** [comment]
@royge when you attempt to send in bulk, do you receive an error message?

**Comment by royge** [comment]
@jamaalscarlett unfortunately, there is not error message return from apns. BTW, I'm using an old version(1.2.1) for some reasons I don't have control :). I haven't tested though if this exists in the latest stable version.
